cmd/controller-manager/main.go

```diff
@@ -134,20 +159,131 @@ func main() {
-	if err != nil {
-		setupLog.Error(err, "unable to get ITS kubeconfig")
+	var itsRestConfig *rest.Config
+	var err2 error
+
+	// Retry getting ITS config for up to 2 minutes to handle temporary unavailability
+	for i := 0; i < 24; i++ { // 24 * 5 seconds = 2 minutes
+		itsRestConfig, itsName, err2 = ctrlutil.GetITSKubeconfig(setupLog, itsName)
+		if err2 == nil {
+			break
+		}
+		setupLog.Info("ITS not available yet, retrying", "attempt", i+1, "error", err2)
+		time.Sleep(5 * time.Second)
+	}
+
+	if err2 != nil {
+		setupLog.Error(err2, "unable to get ITS kubeconfig after retries")
 		os.Exit(1)
 	}
 	setupLog.Info("Got config for ITS", "name", itsName)
 	itsRestConfig = itsClientLimits.LimitConfig(itsRestConfig)
 
+	// Start /readyz endpoint - be resilient to ITS connection issues
+	go func() {
+		http.HandleFunc("/readyz", func(w http.ResponseWriter, r *http.Request) {
+			// Controller-manager is ready if:
+			// 1. It has become leader (readyFlag is true)
+			// 2. OR it's running without leader election (readyFlag is true)
+			// 3. AND it has successfully connected to WDS at least once
+			if readyFlag.Load() {
+				w.WriteHeader(http.StatusOK)
+				w.Write([]byte("ok"))
+			} else {
+				w.WriteHeader(http.StatusServiceUnavailable)
+				w.Write([]byte("not ready"))
+			}
+		})
+		http.ListenAndServe(":8081", nil)
+	}()
+
+	// Start controllers based on leader election setting
+	if enableLeaderElection {
+		setupLog.Info("Starting with leader election enabled")
+		startControllersWithLeaderElection(ctx, setupLog, wdsRestConfig, itsRestConfig, wdsName, itsName, allowedGroupsSet, ctlrsToStart, wdsClientMetrics, itsClientMetrics)
+	} else {
+		setupLog.Info("Starting without leader election")
+		startControllersDirectly(ctx, setupLog, wdsRestConfig, itsRestConfig, wdsName, itsName, allowedGroupsSet, ctlrsToStart, wdsClientMetrics, itsClientMetrics)
+	}
+
+	select {}
+}
+
+// startControllersWithLeaderElection uses client-go leader election instead of controller-runtime
+func startControllersWithLeaderElection(ctx context.Context, setupLog logr.Logger, wdsRestConfig, itsRestConfig *rest.Config, wdsName, itsName string, allowedGroupsSet sets.Set[string], ctlrsToStart sets.Set[string], wdsClientMetrics, itsClientMetrics ksmetrics.ClientMetrics) {
+	// Create WDS client for leader election
+	wdsClient, err := kubernetes.NewForConfig(wdsRestConfig)
+	if err != nil {
+		setupLog.Error(err, "unable to create WDS client for leader election")
+		os.Exit(1)
+	}
+
+	_, err = wdsClient.CoreV1().Namespaces().Get(ctx, lockNamespace, metav1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			setupLog.Info("Creating namespace for leader election", "namespace", lockNamespace)
+			_, err = wdsClient.CoreV1().Namespaces().Create(ctx, &corev1.Namespace{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: lockNamespace,
+				},
+			}, metav1.CreateOptions{FieldManager: fieldManagerName})
+			if err != nil {
```

Comment:
Consider what happens if I create a Deployment with 3 replicas doing this. It is not unlikely that they will concurrently try this Create. One will succeed, the other two fail with an "already exists" error (which you can identify with https://github.com/kubernetes/apimachinery/blob/v0.30.14/pkg/api/errors/errors.go#L538-L542) --- which is not a fatal problem, just a roundabout success.